cmd/contact-auditor/main.go

```diff
@@ -0,0 +1,203 @@
+	"strings"
+	"time"
+
+	"github.com/jmhodges/clock"
+	"github.com/letsencrypt/boulder/cmd"
+	blog "github.com/letsencrypt/boulder/log"
+	"github.com/letsencrypt/boulder/policy"
+)
+
+var queryInterrupted bool
+
+type contactAuditor struct {
+	db     *sql.DB
+	logger blog.Logger
+	clk    clock.Clock
+	grace  time.Duration
+}
+
+type result struct {
+	// Receiver for the `id` column.
+	ID int64
+
+	// Receiver for the `contact` column.
+	Contact  []byte
+	contacts []string
+}
+
+func (r *result) unmarshalContact() error {
+	var contact []string
+	err := json.Unmarshal(r.Contact, &contact)
+	if err != nil {
+		return err
+	}
+	r.contacts = append(r.contacts, contact...)
+	return nil
+}
+
+func (r *result) validateContact() error {
+	// Setup a buffer to store any validation problems we encounter.
+	var probsBuff strings.Builder
+
+	// Helper to write validation problems to our buffer.
+	writeProb := func(contact string, prob string) {
+		// Add validation problem to buffer.
+		probsBuff.WriteString(fmt.Sprintf("[ %q: %q ] ", contact, prob))
+	}
+
+	for _, contact := range r.contacts {
+		if strings.HasPrefix(contact, "mailto:") {
+			err := policy.ValidEmail(strings.TrimPrefix(contact, "mailto:"))
+			if err != nil {
+				writeProb(contact, err.Error())
+			}
+		} else {
+			writeProb(contact, "missing 'mailto:' prefix")
+		}
+	}
+
+	if probsBuff.String() != "" {
+		return errors.New(probsBuff.String())
+	}
+	return nil
+}
+
+// beginAuditQuery executes the audit query and returns a cursor used to
+// stream the results.
+func (c contactAuditor) beginAuditQuery() (*sql.Rows, error) {
+	rows, err := c.db.Query(
+		fmt.Sprintf(
+			"%s '%s';",
+			`SELECT DISTINCT r.id, r.contact
+			FROM registrations AS r
+				INNER JOIN certificates AS c on c.registrationID = r.id
+			WHERE r.contact NOT IN ('[]', 'null')
+				AND c.expires >=`,
+			c.clk.Now().Add(-c.grace).Format("2006-01-02T15:04:05Z")),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("error performing db query: %s", err)
+	}
+	return rows, nil
+}
+
+// run retrieves a cursor from `beginAuditQuery` and then audits the
+// `contact` column of all returned rows for abnormalities or policy
+// violations.
+func (c contactAuditor) run(resChan chan *result) error {
+	c.logger.Infof("Beginning database query")
+	rows, err := c.beginAuditQuery()
+	if err != nil {
+		return err
+	}
+
+	for rows.Next() {
+		var res result
+		err := rows.Scan(&res.ID, &res.Contact)
+		if err != nil {
+			return err
+		}
+
+		err = res.unmarshalContact()
+		if err != nil {
+			c.logger.Errf("Unmarshal failed for ID: %d due to: %s", res.ID, err)
+		}
+
+		err = res.validateContact()
+		if err != nil {
+			c.logger.Errf("Validation failed for ID: %s due to: %s", res.ID, err)
+		}
+
+		// Only used for testing.
+		if resChan != nil {
+			resChan <- &res
+		}
+	}
+	// Ensure the query wasn't interrupted before it could complete.
+	err = rows.Close()
+	if err != nil {
+		// Log an error but continue processing results.
```

Comment:
No further results processing happens after this point -- we've run out of rows, closed the db query, and only then discovered if something went wrong with it. So why not just return an error (which main can handle as non-fatal, if it likes), instead of setting a global boolean?

Reply:
You're right, this was a holdover from a previous iteration.